Adds compiler-rewrite-does-not-apply error condition

matt-y opened the pull request:
Updates relevant call sites previously using the non-specific `GIVE-UP-COMPILATION` to directly signal the new condition when applicable. 

please see rigetti/quilc/issues/46

notmgsk commented on src/compressor/rewriting-rule-data-type.lisp:
```diff
@@ -23,7 +23,7 @@ CONSUMER is a function that consumes a COMPRESSOR-CONTEXT followed by a list of
   (consumer (lambda (context item)
               (declare (ignore context item))
-              (give-up-compilation))
+              (error 'compiler-rewrite-does-not-apply))
```
`(give-up-compilation :because ':rewrite-does-not-apply)`

notmgsk commented on src/compressor/rewriting-rule-data-type.lisp:
```diff
@@ -108,12 +108,12 @@ BODY is a list of forms that, by the end, should construct a list of instruction
 
 (defmacro give-up-compilation-unless (test &body body)
-  "If TEST passes, proceed to execute BODY (and return the value of the final expression). If TEST fails, call GIVE-UP-COMPILATION."
+  "If TEST passes, proceed to execute BODY (and return the value of the final expression). If TEST fails, signal compiler-rewrite-does-not-apply."
```
I'm not convinced this macro shouldn't be `(defmacro give-up-compilation-unless (test because &body body) ...)` to mimic `give-up-compilation`. Likewise this docstring should be reverted + include description of the `because` arg.

stylewarning replied:
Maybe `give-up-rewriting-unless` provided it's _only_ used for rewriting?

notmgsk commented on src/compressor/rewriting-rules.lisp:
```diff
@@ -391,7 +391,7 @@
                  ;; all the bits are nonzero, so we can't reduce: 1111
                  (otherwise
-                  (give-up-compilation)))))))))))
+                  (error 'compiler-rewrite-does-not-apply)))))))))))
```
See above.

stylewarning commented on src/compilation-methods.lisp:
```diff
@@ -38,13 +38,17 @@
+  (:documentation "This is signaled when a rewriting rule cannot be applied to an instruction sequence."))
 
 
 (defun give-up-compilation (&key (because ':unknown))
```
I think we ought to deprecate this function. I added it as a stop-gap measure for portability but that was a mistake I think. Doesn't mean you have to delete it in _this_ PR, but we shouldn't add to it.

(If you want to change the call sites that have something _other_ than `:unknown`, you're welcome to.

stylewarning commented on src/compilation-methods.lisp:
```diff
@@ -38,13 +38,17 @@
   (:documentation "This is signaled when a compiler is technically applicable, but would act as an identity."))
 
+(define-condition compiler-rewrite-does-not-apply (compiler-does-not-apply)
+  ()
```
We could add the rewriting rule to this condition if we wanted to. But not necessary for the PR.

notmgsk replied:
that would be neat

stylewarning commented on src/compressor/rewriting-rule-data-type.lisp:
```diff
@@ -108,12 +108,12 @@ BODY is a list of forms that, by the end, should construct a list of instruction
       ,@body)
      (t
-      (give-up-compilation))))
+      (give-up-compilation :because ':rewrite-does-not-apply))))
```
signal the condition directly

matt-y commented:
@notmgsk @stylewarning Hello I've tidied up the following: 

1. removed my addition to the `GIVE-UP-COMPILATION` method as it appears this is to be deprecated 
2. Renamed the `GIVE-UP-COMPILATION-UNLESS` macro used in the rewrite data type and rules files to be called `GIVE-UP-REWRITING-UNLESS` and updated call sites. 
3. Went back to directly signaling `'compiler-rewrite-does-not-apply` in place of calling `GIVE-UP-COMPILATION` 
4. Updated two other uses of the keyword argument `GIVE-UP-COMPILATION` to directly signal the respective error being specified by the keyword arg

stylewarning commented on src/compilation-methods.lisp:
```diff
@@ -38,13 +38,16 @@
-    (:invalid-domain (error 'compiler-invalid-domain))
-    (:acts-trivially (error 'compiler-acts-trivially))
-    (:unknown        (error 'compiler-does-not-apply))))
+    (:invalid-domain         (error 'compiler-invalid-domain))
```
oof the new spacing

notmgsk commented:
@stylewarning tests pass. Let's go to LGTM.
